test/unit/test_iostream.c

```diff
@@ -0,0 +1,125 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2021, Tarantool AUTHORS, please see AUTHORS file.
+ */
+
+#include <sys/uio.h>
+#include <fcntl.h>
+#include <sys/errno.h>
+
+#include "test_iostream.h"
+#include "trivia/util.h"
+
+static ssize_t
+test_stream_read(struct iostream *io, void *buf, size_t count)
+{
+	struct test_stream_ctx *ctx = io->ctx;
+	io->fd = ctx->rfd;
+	(void)buf;
+	ssize_t nrd = 0;
+	if (ctx->avail_rd == 0)
+		return IOSTREAM_WANT_READ;
+	nrd = MIN(count, ctx->avail_rd);
+	ctx->avail_rd -= nrd;
+	return nrd;
+}
+
+static ssize_t
+test_stream_write(struct iostream *io, const void *buf, size_t count)
+{
+	assert(io->fd >= 0);
+	struct test_stream_ctx *ctx = io->ctx;
+	io->fd = ctx->wfd;
```

Comment:
Resetting `fd` from a callback looks like abusing the `iostream` API. Also this whole `test_iostream` fixture looks redundant. Why not simply create a plain iostream for each test case? Something like this:

```C
int fd[2];
pipe(fd);
// in a new fiber
struct iostream io;
iostream_create(&io, fd[0]);
char c;
coio_read(&io, &c, 1);
iostream_destroy(&io);
// in the main fiber
fiber_wakeup(f);
fiber_sleep(0);
ok(!fiber_is_dead(f));
write(fd[1], ...);
fiber_join(f);
close(fd[0]);
close(fd[1]);
```

Reply:
Yes. The `test_iostream` thingy was only needed before I started using pipes in the test.
Reworked the test accordingly. Please, check out the new version.